### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by andresgarcia29**
- I had some troubles with unit test with sinon!

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/test/*
```
Comment (pateketrueke) [inline .eslintignore:1]: Why? I mean, even tests _should_ follow the same guide-style. IMHO

Comment (andresgarcia29) [reply .eslintignore:1]: I follow the same guide-style in the tests, but I added the .eslintignore cause my editor do not recognize "it" and "describe" and when I see the code with red underline is very stressed for me!

File: Services/Api/Cart/Controller.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        {
+          model: models.OrderProduct,
+          as: 'product_order',
+          include: [{
+            model: models.Product,
+            as: 'product_order',
+          }],
+        },
+      ],
+    }).then((order) => {
```
Comment (pateketrueke) [inline Services/Api/Cart/Controller.js]: WAT! That's the point of using async/away here... I mean, you can omit the `then` calls and such, and treat the return values from promises as is, e.g.

```js
const fn = async () => {
  const x = await Promise.resolve(42);
  console.log(typeof x); // number
  console.log(x); // 42
};
```

Comment (andresgarcia29) [reply Services/Api/Cart/Controller.js]: I have to change all the controllers to grpc, the current controllers are incorrect, but, thanks for the observation!

File: Services/Api/Cart/Controller.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    await models.Order.findById(req.params._id).then((ord) => {
+      if (ord.user !== req.user.id) return res.status(200).send({ message: 'You are not the owner for this order' });
+    }).then(async () => {
+      await models.Product.findById(body.productId).then((product) => {
+        if (!product) return res.status(200).send({ message: "This product doesn't exist" });
+        models.OrderProduct.create(body).then(() => {
+          res.status(200).send({
+            message: 'The product has been added',
+          });
+        }).catch((err) => {
```
Comment (pateketrueke) [inline Services/Api/Cart/Controller.js]: Regardless of the async/await usage you _should_ avoid deep-nesting from `catch()` handlers in your promises.

As long you return a promise-like value you can _catch_ the errors thrown by inner promises, e.g.

```js
Promise.resolve(1)
  .then(() => {
    return Promise.resolve(2)
      .then(() => {
        return Promise.resolve(3)
          .then(() => {
            throw new Error();
          });
      });
  })
  .catch(e => {
    console.log(e.stack);
  });
```

This means you don't need to add a `catch()` handler for each nested promise.

File: Services/Config/Disscount.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      totally += parseInt((object.qty / 2), 10) * object.price;
+      return totally;
+    }
+    totally += (parseInt((object.qty / 2), 10) * object.price) + object.price;
+    return totally;
+  }
+
+  if (object.code === 'tshirt') {
+    if (object.qty >= 3) {
+      totally += object.qty * 19;
```
Comment (pateketrueke) [inline Services/Config/Disscount.js]: - Is not `object.price` available on your input?
- What will happen if the price for t-shirts is changed on the database?

Comment (andresgarcia29) [reply Services/Config/Disscount.js]: Object.price is into the database, in the case of shirt I used a magic number "19" I mean the discount that shirt has when someone buy 3 or more.

File: .eslintignore (hunk 1 above)
Comment (pateketrueke) [reply .eslintignore:1]: Oh, then you need to configure your .eslintrc to recognize _mocha_ calls instead, we'll figure out that later.

File: Services/Config/Disscount.js (hunk 4 above)
Comment (pateketrueke) [reply Services/Config/Disscount.js]: Yeah, but if the price for t-shirt is changed, you'll come back and update that magic number?

I mean, that's not even easier than `(object.price - 1)` right?

Comment (andresgarcia29) [reply Services/Config/Disscount.js]: Yeah! You right, I'll fix it up. Thanks

File: Services/user/Docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  node:
+    build:
+      context: .
+      args:
+        NODE_ENV=development
+    command: ../node_modules/.bin/nodemon --inspect=0.0.0.0:9229
+    ports:
+      "50050:50050"
+    volumes:
+      .:/opt/app:delegated
```
Comment (pateketrueke) [inline Services/user/Docker-compose.yml]: What means this syntax? 🤔

File: Client/order.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
+const grpc = require('grpc');
```
Comment (AalcarazZ-Gen) [inline Client/order.js:1]: Local tests doesn't got to repo

File: Docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  user:
+    build: ./services/user
+    environment:
+      NODE_ENV: development
+    ports:
+      - '50050:50050'
+    depends_on:
+      - db
+  db:
+    image: postgres
```
Comment (AalcarazZ-Gen) [inline Docker-compose.yml:13]: Specify version

File: Services/user/Api/Controllers/order.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      };
+      callback(null, payload);
+    } catch (error) {
+      callback(error);
+    }
+  }
+  async addProduct(call, callback) {
+    try {
+      const getOneOrder = await Order.findById(call.request.idOrder);
+      if (getOneOrder.dataValues.id === call.request.idOrder) {
```
Comment (AalcarazZ-Gen) [inline Services/user/Api/Controllers/order.js:29]: Redundant?

File: Services/user/Api/Controllers/user.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        email: getUser.dataValues.email,
+        firstName: getUser.dataValues.firstName,
+        lastName: getUser.dataValues.lastName,
+      };
+      callback(null, payload);
+    } catch (error) {
+      callback(error);
+    }
+  }
+  async UpdateUser(call, callback) {
```
Comment (AalcarazZ-Gen) [inline Services/user/Api/Controllers/user.js:32]: `updateUser`

File: Services/user/Api/Controllers/user.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    try {
+      await User.upsert(call.request);
```
Comment (AalcarazZ-Gen) [inline Services/user/Api/Controllers/user.js:34]: Change this to update or a find and update.

File: Services/user/Api/Protos/order.proto [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+message ResOrder {
+  int32 id = 1;
+  bool payed = 2;
+  int32 userId = 3;
+  double totally = 4;
+}
+
+
```
Comment (AalcarazZ-Gen) [inline Services/user/Api/Protos/order.proto:32]: Extra blank space

File: Services/user/database/seeders/20180216181357-User.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.bulkInsert('Users', [{
+      firstName: 'Andres',
+      lastName: 'Garcia',
+      password: 'hola123xd',
+      email: 'andres@gmail.com',
+      createdAt: new Date(Date.now()),
```
Comment (AalcarazZ-Gen) [inline Services/user/database/seeders/20180216181357-User.js:8]: `new Date()`

File: Test/unit/Controllers/order.test.js [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+const mocks = require('../../Mocks/controllers/order.controller');
+chai.should();
+
+describe('Test to OrderController', async () => {
+
+  beforeEach(() => {
+    sandBox.restore();
+  });
+
+
```
Comment (AalcarazZ-Gen) [inline Test/unit/Controllers/order.test.js:20]: Extrs line

File: Test/unit/Controllers/order.test.js [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+  it('Should create a new order', async () => {
+    const callback = sandBox.spy();
+    sandBox.stub(Order, 'create').resolves(mocks.orderCreate);
+    return OrderController.createOrder(mocks.orderCreateCall, callback)
+    .should.be.fulfilled
+    .then(() => {
+      assert(callback.calledOnce);
+      expect(callback.args[0][0]).to.be.a('null');
+      expect(callback.args[0][1]).to.be.a('number');
```
Comment (AalcarazZ-Gen) [inline Test/unit/Controllers/order.test.js:29]:
```javascript
callback.calledOnce.should.be.true;
callback.args[0][0].to.be.null;
callback.args[0][1].to.be.a('number');
```

Comment (AalcarazZ-Gen) [reply Test/unit/Controllers/order.test.js:29]: `callback.args[0][0].should.be.null;` => `callback.args[0].should.be.eql(params);`

File: Test/unit/Controllers/order.test.js [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+  });
+
+
+  it('Should create a new order', async () => {
+    const callback = sandBox.spy();
+    sandBox.stub(Order, 'create').resolves(mocks.orderCreate);
+    return OrderController.createOrder(mocks.orderCreateCall, callback)
+    .should.be.fulfilled
+    .then(() => {
+      assert(callback.calledOnce);
```
Comment (AalcarazZ-Gen) [inline Test/unit/Controllers/order.test.js:27]: Add `Order.create` missing assertions

File: Test/unit/Controllers/order.test.js [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+      });
+  });
+
+  it('Should add product to order', async () => {
+    const callback = sandBox.spy();
+    sandBox.stub(Order, 'findById').resolves(mocks.addProductOrderOrder);
+    sandBox.stub(Product, 'findById').resolves(mocks.addProductOrderProduct);
+    return OrderController.addProduct(mocks.addProductRequest, callback)
+    .should.be.fulfilled
+    .then(() => {
```
Comment (AalcarazZ-Gen) [inline Test/unit/Controllers/order.test.js:50]: Add missing `Order.findById` assertions

File: Client/order.js (hunk 6 above)
Comment (andresgarcia29) [reply Client/order.js:1]: Done

File: Docker-compose.yml (hunk 7 above)
Comment (andresgarcia29) [reply Docker-compose.yml:13]: Done with V.10

File: Services/user/Api/Controllers/order.js (hunk 8 above)
Comment (andresgarcia29) [reply Services/user/Api/Controllers/order.js:29]: I couldn't remember why I did it, but currently is fix

File: Services/user/Api/Controllers/user.js (hunk 9 above)
Comment (andresgarcia29) [reply Services/user/Api/Controllers/user.js:32]: Done

File: Services/user/Api/Controllers/user.js (hunk 10 above)
Comment (andresgarcia29) [reply Services/user/Api/Controllers/user.js:34]: I change it and I use update only

File: Test/unit/Controllers/order.test.js (hunk 13 above)
Comment (andresgarcia29) [reply Test/unit/Controllers/order.test.js:20]: Done

File: Test/unit/Controllers/order.test.js (hunk 14 above)
Comment (andresgarcia29) [reply Test/unit/Controllers/order.test.js:29]: Done all the tests

File: Test/unit/Controllers/order.test.js (hunk 16 above)
Comment (andresgarcia29) [reply Test/unit/Controllers/order.test.js:50]: done

File: Test/unit/Controllers/order.test.js (hunk 15 above)
Comment (andresgarcia29) [reply Test/unit/Controllers/order.test.js:27]: done

**Pull request merged**